### PR TITLE
Fix create, push and watch commands on windows

### LINKED ISSUE #641: 'odo watch comp-name' detects and push changes, but nothing gets updated
`odo push' on the other hand works flawlessly.

Below console output for both commands. First `odo watch' executed and nodejs-ex/views/index.html changed in notepad.exe. No changes appears on app page. Then after stopping `odo watch` `odo push` executed and app page gets updated.

**watch**
```
odo watch node-test1 -v
[37mDEBU[0m[0000] Trying to connect to server 10.0.0.48:8443 - %!v(MISSING)
[37mDEBU[0m[0000] Server https://10.0.0.48:8443 is up
[37mDEBU[0m[0000] isLoggedIn err:  <nil>
 output: "developer"
[37mDEBU[0m[0000] Getting BuildConfig: node-test1-a3
[37mDEBU[0m[0000] Source for component node-test1 is file://C:/p/nodejs-ex (local)
[37mDEBU[0m[0000] starting WatchAndPush, path: /p/nodejs-ex, component: node-test1
[37mDEBU[0m[0000] adding watch on path /p/nodejs-ex
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\hooks
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\info
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\logs
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\logs\refs
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\logs\refs\heads
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\logs\refs\remotes
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\logs\refs\remotes\origin
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\03
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\58
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\65
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\78
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\7f
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\81
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\88
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\a4
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\c2
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\ca
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\d4
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\ef
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\f7
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\info
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\objects\pack
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\refs
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\refs\heads
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\refs\remotes
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\refs\remotes\origin
[37mDEBU[0m[0000] ignoring watch for \p\nodejs-ex\.git\refs\tags
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\helm
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\helm\nodejs
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\helm\nodejs\templates
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\openshift
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\openshift\pipeline
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\openshift\templates
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\tests
[37mDEBU[0m[0000] adding watch on path \p\nodejs-ex\views
Waiting for something to change in /p/nodejs-ex
[37mDEBU[0m[0024] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0024] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0024] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0024] adding watch on path \p\nodejs-ex\views\index.html
File \p\nodejs-ex\views\index.html changed
Pushing files...
[37mDEBU[0m[0025] Copying files [\p\nodejs-ex\views\index.html] to pod
[37mDEBU[0m[0025] Waiting for deploymentconfig=node-test1-a3 pod
[37mDEBU[0m[0025] Status of node-test1-a3-1-rxc6n pod is Running
[37mDEBU[0m[0025] Pod node-test1-a3-1-rxc6n is running.
[37mDEBU[0m[0025] Copying to pod node-test1-a3-1-rxc6n
my localpath is /p/nodejs-ex
Please wait, building component....
[33m---> Installing application source ...
[0m[33m---> Installing all dependencies
[0m[33mup to date in 1.072s
[0m[33m---> Building in production mode
[0m[33m---> Pruning the development dependencies
[0m[33mup to date in 0.833s
[0m[33m/opt/app-root/src/.npm is not a mountpoint
[0m[33m---> Cleaning the npm cache /opt/app-root/src/.npm
[0m[33m/tmp is not a mountpoint
[0m[33m---> Cleaning the /tmp/npm-*
[0m[33mrun: stopped
[0m[33mrun: started
[0mWaiting for something to change in /p/nodejs-ex
[37mDEBU[0m[0176] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0176] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0176] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0176] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0176] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0176] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0176] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0176] adding watch on path \p\nodejs-ex\views\index.html
File \p\nodejs-ex\views\index.html changed
Pushing files...
[37mDEBU[0m[0177] Copying files [\p\nodejs-ex\views\index.html] to pod
[37mDEBU[0m[0177] Waiting for deploymentconfig=node-test1-a3 pod
[37mDEBU[0m[0177] Status of node-test1-a3-1-rxc6n pod is Running
[37mDEBU[0m[0177] Pod node-test1-a3-1-rxc6n is running.
[37mDEBU[0m[0177] Copying to pod node-test1-a3-1-rxc6n
my localpath is /p/nodejs-ex
Please wait, building component....
[33m---> Installing application source ...
[0m[33m---> Installing all dependencies
[0m[33mup to date in 0.96s
[0m[33m---> Building in production mode
[0m[33m---> Pruning the development dependencies
[0m[33mup to date in 0.883s
[0m[33m/opt/app-root/src/.npm is not a mountpoint
[0m[33m---> Cleaning the npm cache /opt/app-root/src/.npm
[0m[33m/tmp is not a mountpoint
[0m[33m---> Cleaning the /tmp/npm-*
[0m[33mrun: stopped
[0m[33mrun: started
[0mWaiting for something to change in /p/nodejs-ex
[37mDEBU[0m[0251] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0251] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0251] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0251] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0251] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0251] adding watch on path \p\nodejs-ex\views\index.html
[37mDEBU[0m[0251] filesystem watch event: "\\p\\nodejs-ex\\views\\index.html": WRITE
[37mDEBU[0m[0251] adding watch on path \p\nodejs-ex\views\index.html
File \p\nodejs-ex\views\index.html changed
Pushing files...
[37mDEBU[0m[0253] Copying files [\p\nodejs-ex\views\index.html] to pod
[37mDEBU[0m[0253] Waiting for deploymentconfig=node-test1-a3 pod
[37mDEBU[0m[0253] Status of node-test1-a3-1-rxc6n pod is Running
[37mDEBU[0m[0253] Pod node-test1-a3-1-rxc6n is running.
[37mDEBU[0m[0253] Copying to pod node-test1-a3-1-rxc6n
my localpath is /p/nodejs-ex
Please wait, building component....
[33m---> Installing application source ...
[0m[33m---> Installing all dependencies
[0m[33mup to date in 0.96s
[0m[33m---> Building in production mode
[0m[33m---> Pruning the development dependencies
[0m[33mup to date in 0.994s
[0m[33m/opt/app-root/src/.npm is not a mountpoint
[0m[33m---> Cleaning the npm cache /opt/app-root/src/.npm
[0m[33m/tmp is not a mountpoint
[0m[33m---> Cleaning the /tmp/npm-*
[0m[33mrun: stopped
[0m[33mrun: started
[0mWaiting for something to change in /p/nodejs-ex
```
**push**
```
c:\p\nodejs-ex>odo push
Pushing changes to component: node-test1
my localpath is /p/nodejs-ex
Please wait, building component....
[33m---> Installing application source ...
[0m[33m---> Installing all dependencies
[0m[33mup to date in 1.372s
[0m[33m---> Building in production mode
[0m[33m---> Pruning the development dependencies
[0m[33mup to date in 1.17s
[0m[33m/opt/app-root/src/.npm is not a mountpoint
[0m[33m---> Cleaning the npm cache /opt/app-root/src/.npm
[0m[33m/tmp is not a mountpoint
[0m[33m---> Cleaning the /tmp/npm-*
[0m[33mrun: stopped
[0m[33mrun: started
[0mchanges successfully pushed to component: node-test1
```

### PR DESCRIPTION
**Pull request by dgolovin**
Fix #641 and #635

### CODE REVIEW

File: pkg/component/component.go [hunk 1]
```diff
@@ -131,6 +132,25 @@ func CreateFromPath(client *occlient.Client, name string, componentImageType str
 	return nil
 }
 
+// Reads file path form URL file:///C:/path/to/file to C:\path\to\file
+func ReadFilePath(u *url.URL) string {
```
Comment (surajnarwade) [inline pkg/component/component.go]: this functions  `ReadFilePath` and `GenFilePath` should go in `pkg/util/util.go`

Comment (surajnarwade) [inline pkg/component/component.go]: this functions `ReadFilePath` and `GenFilePath` should go in `pkg/util/util.go`

**Comment by surajnarwade** [comment]
other than one comment, everything looks good to me, worked on my linux machine
Thanks @dgolovin :)

**Comment by dgolovin** [comment]
@surajnarwade thanks for the review. I've moved those methods to `util` package.

**Comment by dgolovin** [comment]
@kadel could you review it or assign someone?

**Comment by surajnarwade** [comment]
cc @anmolbabu @mik-dass @syamgk please review this PR

**Comment by mik-dass** [comment]
@dgolovin @surajnarwade I will test this on a windows machine

File: pkg/occlient/occlient.go [hunk 2]
```diff
@@ -1647,7 +1647,7 @@ func tar(tw *taro.Writer, fileName string, destFile string) error {
 	}
 	splitFileName := strings.Split(fileName, destFile)[1]
 
-	hdr.Name = destFile + splitFileName
+	hdr.Name = destFile + strings.Replace(splitFileName, "\\", "/", -1)
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: Mind adding a small comment that this is intended for Windows / Mac OS / Linux support?

File: pkg/util/util.go [hunk 3]
```diff
@@ -59,3 +62,22 @@ func NamespaceOpenShiftObject(componentName string, applicationName string) (str
 	// Return the hyphenated namespaced name
 	return fmt.Sprintf("%s-%s", componentName, applicationName), nil
 }
+
+// Reads file path form URL file:///C:/path/to/file to C:\path\to\file
+func ReadFilePath(u *url.URL) string {
```
Comment (cdrage) [inline pkg/util/util.go]: We should add tests for this.. but we can ignore this for now just so we can fix the commands on Windows for the time being. Can you add a TODO: Write tests comment?

Comment (dgolovin) [reply pkg/util/util.go]: sure, I'll look if I can add tests, to make PR checks back to green, if I fail, then adding `TODO` is my backup plan.

**Comment by anmolbabu** [comment]
@dgolovin does this PR fix https://github.com/redhat-developer/odo/issues/639 also ?

**Comment by dgolovin** [comment]
@anmolbabu I think #639 is a different problem and this fix is not going to help.

**Comment by dgolovin** [comment]
@cdrage I've added tests and comment, why replacing '\' to '/' is required.

**Comment by dgolovin** [comment]
I've squashed all commits in one and fixed some typos in test messages. 
@cdrage @anmolbabu @mik-dass @surajnarwade  I have no rights to push, could you release it to the master branch?

File: pkg/util/util.go [hunk 4]
```diff
@@ -59,3 +61,24 @@ func NamespaceOpenShiftObject(componentName string, applicationName string) (str
 	// Return the hyphenated namespaced name
 	return fmt.Sprintf("%s-%s", componentName, applicationName), nil
 }
+
+const WIN = "windows"
+
+// Reads file path form URL file:///C:/path/to/file to C:\path\to\file
```
Comment (mik-dass) [inline pkg/util/util.go]: Shouldn't it be C:\\path\\to\\file?

Comment (mik-dass) [inline pkg/util/util.go]: Shouldn't it be C:\\\path\\\to\\\file ?

**Comment by dgolovin** [comment]
@mik-dass fixed. Thanks!

**Comment by cdrage** [comment]
Unfortunately this needs a re-based as we merged in some pretty big PRs. The code LGTM however! So after re-base, feel free to merge.

**Comment by dgolovin** [comment]
@cdrage rebased on top of master.

**Pull request merged**